wrapanapi/base/system.py

```diff
@@ -0,0 +1,166 @@
+
+    Interface notes:
+
+    * Initializers of subclasses must support \*\*kwargs in their
+      signtures
+
+    """
+    __metaclass__ = ABCMeta
+    STEADY_WAIT_MINS = 3
+
+    @classmethod
+    @abstractproperty
+    def can_suspend(cls):
+        """Indicates whether this system can suspend VM's/instances."""
+        raise NotImplementedError
+
+    @classmethod
+    @abstractproperty
+    def can_pause(cls):
+        """Indicates whether this system can pause VM's/instances."""
+        raise NotImplementedError
+
+    @classmethod
+    @abstractproperty
+    def steady_wait_time(cls):
+        """Returns default secs to wait for VM to move into a steady state."""
+        return 120
+
+    @abstractmethod
+    def list_vms(self, **kwargs):
+        """Returns a list of VM objects
+
+        Returns: list of BaseVM
+        """
+        raise NotImplementedError('list_vm not implemented.')
+
+    @abstractmethod
+    def create_vm(self, vm_name, *args, **kwargs):
+        """Creates a vm.
+
+        Args:
+            vm_name: name of the vm to be created
+        Returns: BaseVM if creation successful
+        """
+        raise NotImplementedError('create_vm not implemented.')
+
+    @abstractmethod
+    def list_templates(self):
+        """Returns a list of templates/images.
+
+        Returns: list of template/image names
+        """
+        raise NotImplementedError('list_template not implemented.')
+
+    @abstractmethod
+    def list_flavors(self):
+        """Returns a list of flavors.
+
+        Only valid for OpenStack and Amazon
+
+        Returns: list of flavor names
+        """
+        raise NotImplementedError('list_flavor not implemented.')
+
+    def list_networks(self):
+        """Returns a list of networks.
+
+        Only valid for OpenStack
+
+        Returns: list of network names
+        """
+        raise NotImplementedError('list_network not implemented.')
+
+    @abstractmethod
+    def info(self):
+        """Returns basic information about the mgmt system.
+
+        Returns: string representation of name/version of mgmt system.
+        """
+        raise NotImplementedError('info not implemented.')
+
+    @abstractmethod
+    def disconnect(self):
+        """Disconnects the API from mgmt system"""
+        raise NotImplementedError('disconnect not implemented.')
+
+    @abstractmethod
+    def remove_host_from_cluster(self, hostname):
+        """remove a host from it's cluster
+
+        :param hostname: The hostname of the system
+        :type  hostname: str
+        :return: True if successful, False if failed
+        :rtype: boolean
+
+        """
+        raise NotImplementedError('remove_host_from_cluster not implemented.')
+
+    @contextmanager
+    def steady_wait(self, minutes):
+        """Overrides original STEADY_WAIT_MINS variable in the object.
+
+        This is useful eg. when creating templates in RHEV as it has long Image Locked period
```

Comment:
I am presuming that this ```steady_wait``` would not be a publicly accessible method and that the wrapping of methods with longer wait times is something that would be done inside the wrapanapi package. I'm on the fence as to whether I like this. Why doesn't RHV override this value at the class level? Is this for a user to specify an artificially longer amount of time based on their knowledge of the size of the template/state of the environment?

Reply:
Hm yeah sorry, this is a method that I've copied over from the old code without analyzing closely yet ;) I don't even see the "steady_wait" context manager being used in integration_tests so I think this should be safe to remove?

My plan is actually to replace the "STEADY_WAIT_MINS" class variable with an abstract class property (steady_wait_time). That way each provider type is required to define its default steady wait time. 

Then wait_vm_steady() will have an arg to override that time on a call-by-call basis if you want.